esrally/storage/_client.py

```diff
@@ -0,0 +1,220 @@
+        random: Random | None = None,
+        max_connections: int = MAX_CONNECTIONS,
+    ):
+        if random is None:
+            random = Random(time.time())
+        self._adapters: AdapterRegistry = adapters
+        self._cached_heads: dict[str, tuple[Head | Exception, float]] = {}
+        self._connections: dict[str, WaitGroup] = defaultdict(lambda: WaitGroup(max_count=max_connections))
+        self._lock = threading.Lock()
+        self._mirrors: MirrorList = mirrors
+        self._random: Random = random
+        self._stats: dict[str, deque[ServerStats]] = defaultdict(lambda: deque(maxlen=100))
+
+    def head(self, url: str, ttl: float | None = None) -> Head:
+        """It gets remote file headers."""
+
+        start_time = time.monotonic()
+        if ttl is not None:
+            # when time-to-leave is given, it looks up for pre-cached head first
+            try:
+                value, last_time = self._cached_heads[url]
+            except KeyError:
+                pass
+            else:
+                if start_time <= last_time + ttl:
+                    # cached value or error is enough recent to be used.
+                    return _head_or_raise(value)
+
+        adapter = self._adapters.get(url)
+        try:
+            value = adapter.head(url)
+        except Exception as ex:
+            value = ex
+        end_time = time.monotonic()
+
+        with self._lock:
+            # It records per-server time statistics.
+            self._stats[_server_key(url)].append(ServerStats(url, start_time, end_time))
+            # The cached value could be an exception, or a head. In this way it will not retry previously failed
+            # urls until the TTL expires.
+            self._cached_heads[url] = value, start_time
+
+        return _head_or_raise(value)
+
+    def resolve(self, url: str, content_length: int | None = None, accept_ranges: bool = False, ttl: float = 60.0) -> Iterator[Head]:
+        """It looks up mirror list for given URL and yield mirror heads.
+        :param url:
+        :param content_length: if not none it will filter out mirrors with a wrong content length.
+        :param accept_ranges: if True it will filter out mirrors that are not supporting ranges.
+        :param ttl: the time to live value (in seconds) to use for cached heads retrieval.
+        :return: iterator over mirror URLs
+        """
+
+        mirror_urls = list(self._mirrors.resolve(url))
+        if not mirror_urls:
+            yield self.head(url, ttl=ttl)
+            return
+
+        # It shuffles mirror URLs in the hope two threads will try different mirrors for the same URL.
+        self._random.shuffle(mirror_urls)
+
+        # It uses measured latencies and the number of connections to affect the order of the mirrors.
+        weights = {u: self._average_latency(u) * self._random.uniform(1.0, self._server_connections(url).count) for u in mirror_urls}
+        # LOG.debug("weights: %s", weights)
```

Comment:
Hmm, these is a low enough log level, why not leave them in? Or just change them to `LOG.trace`?